// node-red-contrib-ovia-os/os/os.js
/**
 * Copyright 2016 Argonne National Laboratory.
 *
 * Licensed under the BSD 3-Clause License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * ibuted under the License is ibuted on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

module.exports = function(RED) {
    var settings = RED.settings;
    var os = require('os');
    var df = require('node-df');
    var fs = require('fs');
    var Promise = require('promise');

    function OS(n) {
        RED.nodes.createNode(this,n);
        var node = this;
        this.name = n.name;

        node.on("input", function(msg) {
            msg.payload = {
                hostname: os.hostname(),
                type: os.type(),
                platform: os.platform(),
                arch: os.arch(),
                release: os.release(),
                endianness: os.endianness(),
                tmpdir: os.tmpdir()
            };
            node.send(msg);
        });

    }

    RED.nodes.registerType("OS",OS);

    function Drives(n) {
        RED.nodes.createNode(this,n);
        var node = this;
        this.name = n.name;

        node.on("input", function(msg) {
            df(function (error, response) {
                if (error) { throw error; }
                msg.payload = response;
                node.send(msg);
            });
        });

    }

    RED.nodes.registerType("Drives",Drives);

    function Uptime(n) {
        RED.nodes.createNode(this,n);
        var node = this;
        this.name = n.name;

        node.on("input", function(msg) {
            msg.payload = {uptime: os.uptime()};
            node.send(msg);
        });
    }

    RED.nodes.registerType("Uptime",Uptime);

    function CPUs(n) {
        RED.nodes.createNode(this,n);
        var node = this;
        this.name = n.name;

        node.on("input", function(msg) {
            msg.payload = {cpus: os.cpus()};
            node.send(msg);
        });
    }

    RED.nodes.registerType("CPUs",CPUs);

    function Loadavg(n) {
        RED.nodes.createNode(this,n);
        var node = this;
        this.name = n.name;

        node.on("input", function(msg) {
            msg.payload = {loadavg: os.loadavg()};
            node.send(msg);
        });
    }

    RED.nodes.registerType("Loadavg",Loadavg);

    function Memory(n) {
        RED.nodes.createNode(this,n);
        var node = this;
        this.name = n.name;

        node.on("input", function(msg) {
            const path = '/proc/meminfo';
            const dataPromise = fetchData(path)

            let hash = {};
            dataPromise.then(data => {

                memInfoLines = data.toString().split(/\n/g);
                for (var i = 0; i < memInfoLines.length; i++) {
                    singleMemInfoLine = memInfoLines[i].split(':')
                    if (singleMemInfoLine
                        && singleMemInfoLine[0]
                        && singleMemInfoLine[1]
                        && singleMemInfoLine[1].toString().includes("kB")
                    ) {
                        singleMemInfoLine[1] = singleMemInfoLine[1].replace(/kB/, '').trim();
                        var itemType = singleMemInfoLine[0];
                        var itemVal = singleMemInfoLine[1];
                        hash[itemType] = itemVal;
                    }
                }
                msg.payload = hash
                node.send(msg);

            }).catch(function(err) {
                node.error(err);
            });
        });
    }

    function fetchData(path) {
        return new Promise(function(resolve, reject) {
            fs.readFile(path, 'utf8', (err, data) => {
                if(err) reject(err);
                resolve(data)
            });
        });
    }

    RED.nodes.registerType("Memory",Memory);

    function NetworkIntf(n) {
        RED.nodes.createNode(this,n);
        var node = this;
        this.name = n.name;

        node.on("input", function(msg) {
            msg.payload = {networkInterfaces: os.networkInterfaces()};
            node.send(msg);
        });
    }

    RED.nodes.registerType("NetworkIntf",NetworkIntf);
}
